### DISCUSSION TITLE
I want to make a leptos example using this crate, but I'm a little confused on how best use it?

### DISCUSSION THREAD

**Discussion by sjud** [sebadob/rauthy#276] [Q&A]
It looks like, I would run it alongside the server and then use the rauthy client to make calls to it? And it would act like google SSO acts but instead it is my own SSO. So if I deployed this next to my server, I'd basically be the SSO auth provider?
How best would I use this in a leptos example? Thanks for your work on this project!!

**Comment by sebadob** [comment]
> It looks like, I would run it alongside the server and then use the rauthy client to make calls to it? And it would act like google SSO acts but instead it is my own SSO. So if I deployed this next to my server, I'd basically be the SSO auth provider?

correct

A leptos example, it depends if your leptos app is running with SSR or CSR. If it's SSR, then you could either do it like in one of the examples for the `rauthy-client` https://github.com/sebadob/rauthy/tree/main/rauthy-client/examples. It depends on your leptos setup of course, but for ease of use, you could just return the very minimal JS callback answer like in the examples. That's what I do in my apps as well, because the page is empty anyway. If you want to use leptos for the `/callback` as well and show some UI, you actually may need to expand your leptos setup a bit and handle the route requests for yourself. I am doing this too for all apps for different reasons. The main benefit is, that you can build up the context before actually rendering leptos itself.

As an example for an axum backend, you can do the following (since leptos v0.6 you don't need to specify the sever fn handler like in this screen shot any more):

![grafik](https://github.com/sebadob/rauthy/assets/62848396/5fa0ae32-104e-4d3a-8b2b-4bfa87233f5e)

Next, you can handle all the leptos routes yourself and have full control over it:

![grafik](https://github.com/sebadob/rauthy/assets/62848396/c5c5848f-0b26-4826-8ad4-26e1cd0b72bc)

These example screenshots are taken from a small app I am working on right now, so they include more stuff than you would need (maybe).

You can take a peek at the client's cookies and HTTP header before even starting to render the leptos part. This way you can extract a session cookie or maybe an `Authorization` header beforehand, validate it, or redirect the user to a login, if he has no access to the requested route.

In the end, the whole thing would roughly look like that:

- Take a look at the client's access to a given route in the custom route handler via cookies, headers, ...
- If The client has a valid login, session, token, ..., just go on like normal. Otherwise, send an HTTP 301 and redirect to the login endpoint from Rauthy. This will be the `/auth/v1/authroize` with the proper values.
- Upon successful login with Rauthy, the user will be redirected to your callback URL, which you configure for the given client you are using in Rauthy. A minimal example how these could look can be found in the rauthy-client examples https://github.com/sebadob/rauthy/blob/main/rauthy-client/examples/actix-web/src/routes.rs.
- On the callback endpoint of your app, you will check the request and use the `code` from it to fetch the tokens for the user.
- At this point, it highly depends on your app, how you want to proceed. You could either go on with the user just using the `access_token` for authentication, or create a session. Both approaches have advantages and disadvantages of course.

Something else you could do is to take a look at [leptos_oidc](https://crates.io/crates/leptos_oidc). I have never used it so far, because I use the rauthy-client everywhere, but this has been [tested and is working with Rauthy](https://gitlab.com/kerkmann/leptos_oidc/-/blob/main/docs/backends/rauthy.md) as well. There are even examples.

If your app is using CSR, the whole thing would get a bit easier. Let me know if this is the case and I can give some guidance on that as well. I am just assuming you use leptos in SSR mode, which makes the most sense.

**Reply by TheToddmeister** [reply]
Hey!
I looked up this question because i was interested in CSR!
I'm already working on a setup, but if you have any insights on that i would love it!
*if i find this further down, ignore this

**Reply by TheToddmeister** [reply]
Ok, correction. I am actually using the leptos csr example along with the server_fn.
I think the template is called csr_todo_server_fn or something. But since i'm kinda using server fns already i figured i could use the sso example.

**Reply by sebadob** [reply]
If you have any backend, you should handle the login logic there and not inside the browser. It makes your life a lot easier and you have more control, and the option to use confidential clients.

**Reply by TheToddmeister** [reply]
i intend to use a staggered token logic where i use Rauthy to fetch a token from my application.
That token becomes the read token and a write token while Rauthy handles the authentication refresh logic.
The idea is to have a longer term read capability and to reduce the exposure of the authentication token
Good idea or bad?

**Reply by sebadob** [reply]
Hmm I would probably not make it that complicated and use short living tokens + `refresh_token`s. The biggest issue is, that it's impossible to revoke a stateless token by design. But, for `refresh_token`s, that's totally possible, because each refresh triggers additional verification inside Rauthy and it makes sure, that all conditions are met and it's okay to refresh. Depending on your traffic and resource usage, you could even go that far down and making `access_token`s valid for only 5 minutes. Each refresh is "costly" of course, but that would make it almost negligible if a token gets stolen.

Refresh tokens are very powerful and it's not too costly to refresh, as long as you don't do it every 10 seconds for each client of course. If also depends a lot on how your users access your application. If it's via a desktop or mobile app, I e.g. have no issues with longer lived tokens like maybe even multiple days, because it's very unlikely that these get stolen.

If you want the max security, the best thing you can do is to use tokens for session creation and manage a real, proper session on your client, which you can revoke at any given time immediately. Otherwise, tokens with a lifetime between 5 and 15 minutes is pretty safe.

Another approach to get longer read than write would be to set an encrypted / signed cookie on your client app with an embedded expiry for read-only, even without a valid token, and refresh this expiry whenever a new, valid `access_token` has been created.

**Reply by TheToddmeister** [reply]
Thanks for the input!
I think i have settled on a solution, but i'll explain a little more for the fun of it.
I have an application with low read security. It is mostly that it reads a lot and heavy so i want to limit the request burden with authentication.
It also hosts some GDPR info that i dont want to persist in the application.
The application is generally used rarely and i dont want users to log in for a read session.

Most users are never going to write and less than 1% of the time is going to be spent writing.
Write abuse can be very very bad so i have settled on this:
Write and read
 
My intent is mostly to reduce unwanted traffic, so i think it is pretty safe to just implement some anti-abuse logic to ban abusive token-users. It is more critical that the write system is tight.

I intend to have write and user info tied to the rauthy session, but i'll allow a read token to persist on the device long term so that the read capabilities of the application persits beyond the rauty session.

**Reply by sebadob** [reply]
> My intent is mostly to reduce unwanted traffic

Oh yeah, I did a lot in that direction as well in the last year. Some applications, that are used rarely, were hit with ~95% bot and API scanner traffic. It has gotten pretty insane, especially since the many AI bots came up as well, and they don't respect any headers, robots.txt, no nothing, and they sometimes fire like 30req/s. 

Maybe you are interested in this - requesting challenges is the best option currently imho. This blocks 100% of bots in all my tests over the last year. You can do this super easily on your own and you don't need any Cloudflare or others. For instance, you could just use [spow](https://github.com/sebadob/spow) and set a cookie that lives for a few hours, after a client has solved a challenge successfully.

**Reply by TheToddmeister** [reply]
Holy mother of something something!
Are you the most  prolific open source programmer or what? 
That is a really elegant solution for what i want to do,. I might just add this to a lot of the endpoints as well to prevent scrapers from dragging down the website.

**Reply by sebadob** [reply]
> Are you the most prolific open source programmer or what?

Well, `spow` is used in Rauthy as well during either an open user registration or during login, to avoid exactly these issues. It's hands down the best counter for brute-force I can think of. Normal users don't care if their device solves a puzzle for like half a second, but bots and brute-force attacks become way too expensive to run at that point.
Only when you want your site to be indexed by search engines, this will not work of course, because they would be blocked too, but I guess that's not the case for you.

Rauthy only requests this in very few situations, because it has the session cookie otherwise. If you want to protect your whole app, the easiest setup is to have a middleware with some path exceptions, which checks for this all the time. I wanted to create a tiny crate which does exactly that, but I am a bit short on time. However, it's pretty simple to set up on your own. The middleware either redirects the user to another endpoint which solves the PoW, or returns a tiny piece of HTML and Javascript directly, which triggers a site reload on success. This is probably the cleanest solution.

**Reply by TheToddmeister** [reply]
i Have another question.
How do i programmatically add custom_claims programatically, eg without using the UI?
I've been looking around, but it doesnt seem obvious

**Reply by TheToddmeister** [reply]
I found bootstrap. Seems cumbersome, but that seems to be the way

**Reply by sebadob** [reply]
API Key with access to `UserAttributes` + `create` ->  `POST /auth/v1/users/attr` -> https://iam.sebadob.dev/auth/v1/docs/#/users/post_cust_attr

**Comment by sebadob** [comment]
Another much simpler approach to the one above, but of course with a different UX as well, would be to simply not use any custom routes extractor and return 301's from any protected endpoints, if the user is not logged in.  
Meaning, if you access `/protected` from the leptos app and the user did not provide a valid token / session, simply return HTTP 301 and redirect to the login endpoint.  
This would completely get rid of the custom routes handler, if you don't like the complexity. Then you only need to care about the callback endpoint. The rest will work as mentioned.

**Comment by sjud** [comment]
Incredible! Very informative and thorough. I'll be working on this in the next couple of weeks (hopefully) and will follow up here (if this is a good place to do so).
Thank you!

**Comment by sebadob** [comment]
> if this is a good place to do so

If you are just asking for advice or something, starting a [discussion](https://github.com/sebadob/rauthy/discussions) would be cleaner. But if you have an issue, then this is the right place, yes.

So I guess this issue can be closed for now, since there just is none so far?

**Comment by erlend-sh** [comment]
Looks like you may as well move this from Issues to Discussion :) GitHub lets you do that.

**Comment by sebadob** [comment]
Thanks  @erlend-sh . Didn't know github can do this. :)

**Comment by sjud** [comment]
Oh thanks! :)

**Comment by sebadob** [comment]
I got an E-Mail because of a post from you @sjud asking about upstream SSO providers, but I cannot see the post here.  
Anyway, yes I am planning to support upstream auth providers. There is even an issue about it already #166 .

The title may not be totally clear, but when I will implement this issue, it will be a generic support for any upstream auth provider.

**Comment by sjud** [comment]
> I got an E-Mail because of a post from you @sjud asking about upstream SSO providers, but I cannot see the post here. Anyway, yes I am planning to support upstream auth providers. There is even an issue about it already #166 .
> 
> The title may not be totally clear, but when I will implement this issue, it will be a generic support for any upstream auth provider.

Yes, I did post that. But then as I dug through the information available I found an answer to my question so I went and I deleted my question (I didn't want anyone to waste their time on it) little did I know you got an email about it. :)

Regarding a generic implementation, it seems like various popular SSO providers have slightly different processes which might make a fully generic implementation a pain. It might be beneficial for there to be some specific implementations for upstream auth providers, such the big five or six google, facebook, github, apple, microsoft, etc.

This would satisfy some aspiring rauthy users, providing their needed functionality for their onboarding process while also possibly laying the groundwork for a generic implementation later. Having a handful of specific implementations make the design of the generic much easier in my experience.

I've been reading the rauthy codebase, and I find it very accessible and I appreciate the way you've written it. You're responsiveness and willingness to share technical details is very inspiring and I'd like to contribute to the rauthy code base.

May I write a specific implementation for one such upstream auth provider and open a pull request for it? From there we can tweak details and I could hammer out all the popular ones and lay the groundwork for a future generic implementation. I'd love to get involved in this project! Please let me know what you think, thank you.

**Reply by sebadob** [reply]
> I've been reading the rauthy codebase, and I find it very accessible and I appreciate the way you've written it. You're responsiveness and willingness to share technical details is very inspiring and I'd like to contribute to the rauthy code base.

Thanks a lot! :)

With generic I just mean, that any provider should work with the same setup and options.  
I expect that all of them will work with auto-configuration anyway. Meaning, my idea is to only provide the issuer URL and Rauthy will do the lookup for the `openid-configuration` and do the rest.  
To get some templates, the idea for now is to just have a dropdown in the Admin UI, which will fill the config inputs automatically.

If you just give me a few more days, I will lay down the groundwork for this feature. I already have an idea about it, how we can do this dynamically and support any amount of providers at the same time.

**Reply by sjud** [reply]
For sure! No rush. :) 
Can I do some work on the documentation? I think I could be helpful there in presenting some of the concepts to users of rauthy. I think some people will want to use rauthy as a simpler alternative to keycloak and I think if I were to explain the concepts of realms and users, etc, and how to get the service up in running in a few typical self-hosted environments that would speed up early adoption. Also you should snag the domain name for rauthy
https://www.namecheap.com/domains/registration/results/?domain=rauthy
:)

**Reply by sebadob** [reply]
Oh yes, some book updates would be very nice.  
The last new features are mostly missing in the book. Some update with screenshots would make things a lot easier for new users as well.  
I also started creating a helm chart quite a while ago, but then I first implemented new features which would make the chart outdated again, so I postponed this. But now, we are almost there feature wise and not far from v 1.0.0. This will help new users a lot as well most probably.

The concept of realms in Keycloak is actually nice and mandatory for them, because they have the heavy weight JVM under the hood. If you need multiple realms with Rauthy, just depoly another instance, and you would get the maximum amount of flexibility and it does not really use that much additional resources anyway. That's why I did not even bother implementing something like this.

I already do have a few domain names. I am not sure yet, if rauthy will get its own or if it will be combined with a few other of my projects like [Nioca](https://github.com/sebadob/nioca) (which I will bring forward as soon as Rauthy has the SSO upstream providers feature implemented). But yes, domains are not that expensive anyway, so why not. :) Makes sense for v1.0.0 at least at some point.

**Comment by sjud** [comment]
Oops I deleted another comment again. Sorry!

**Reply by sebadob** [reply]
> Have you done any quantitative comparisons of rauthy vs keycloak/Ory with regards to memory footprint, start up times, req handling speed and through put? One problem with both of those services is that I expect their usage of garbage collected languages to give a "spiky" performance, Discord engineering blog explained that as their reason from moving away from Go to Rust for some back end services. Something like an authorization server which is so essential to the user experience should really be written in a low level langauge and I expect this project to get a lot of attention! :)
I'm finishing the leptos rauthy example and will use what I learn there to begin work on a new users guide. Thanks!

Not yet. This is one of the very last steps before v1.0.0 - benchmarks and performance tuning. There will not be too much to do, because I did a lot of this stuff already. The whole code base has been almost fully rewritten ~3.5 times by now, with quite a bit of optimizations.

Startup time of any Java application is not even worth mentioning when you have Rust or Go. The only thing that makes Rauthy start not as fast as it could are quality of life implementations, additional checks that a user cannot accidentially deploy a mismatching version (database version vs app version checks for instance) and such things. 

Memory wise, Rauthy is, again, not as efficient as it could be, for a good reason.  
Everything you need during normal login flows is actually cached almost for forever (12h by default). The cache is being updated preemptively when data is created or updated, which means it will already exist in the cache before the first GET comes in. Cache invalidation is handled manually for each entry as well. This means there is quite a bit of overhead to manage, but instead any data is just there in no time. The main reason why I did this a few years ago was to counter slow disk access (Raspberry Pi with SD cards for instance). Also, when you use a hosted Postgres which is a bit further away, you will not get into trouble with latency.

I have not done a single benchmark yet, but I expect Rauthy to be pretty good throughput wise, because
1. Rust + Axum is just super fast
2. Everything will be kept in a local cache copy all the time
But the performance here will be really good for Ory as well, I guess (never used it). Keycloak is written in Java, but with Quarkus under the hood, its pretty fast as well. Actually, all of the examples will provide good enough speed for the real world. When you get into millions of users, things will start to make a difference. The big thing is behind the scenes. Java will simply need more resources and is more expensive to run in production than Go, and then Rust. If you throw enough resources at it, it will perform with (mostly) any user count.  
That is something that Rauthy still has a TODO for. The Admin UI users component uses client side pagination currently. This means at some point, the first fetch of the users will take longer than with smaller instances. As long as your dataset is small enough, I always prefer client side pagination, because the UI just feels that much more snappy. Rauthy will need some alternative component there which will be loaded if for instance user count will reach 50k (or whatever makes sense -> benchmarking TODO).

The spiky performance is true because of GC for sure, yes. But I don't think that you will notice it that much with an SSO provider, because you don't login that often than you write messages in discord for instance.

Its quite hard to compare metrics like these between applications, when they simply do different things under the hood.  
But I am curious how Rauthy will perform when the first benchmarks are being done. There are a few TODOs in the code currently at places where I already know that they can be optimized, but optimization does not make any sense when you don't have benchmarks to verify them.
